python/graphstorm/gconstruct/transform.py

```diff
@@ -935,6 +910,149 @@ def call(self, feats):
+        """
+        return self._target_etype
+
+    @property
+    def neg_ntype(self):
+        """ Return the node type of hard negatives
+        """
+        return self._target_ntype
+
+    def set_id_maps(self, id_maps):
+        """ Set ID mapping for converting raw string ID to Graph ID
+        """
+        assert self._target_ntype is not None, \
+            "The target node type should be set, it can be the source node type " \
+            "or the destination node type depending on the hard negative case."
+        assert self._target_ntype in id_maps, \
+            f"The nid mapping should have the mapping for {self._target_ntype}. " \
+            f"But only has {id_maps.keys()}"
+        self._nid_map = id_maps
+
+    def pre_process(self, feats):
+        """ Pre-process input feats
+
+            Not all the edges have the same number of hard negatives.
+            Thus we need to know the maxmun number of hard negatives first.
+
+        Parameters
+        ----------
+        feats:
+            feats to be processed
+
+        Return:
+            dict: {feature_name: feats_statistics}
+        """
+        assert isinstance(feats, (np.ndarray, ExtMemArrayWrapper)), \
+            "Feature of HardEdgeNegativeTransform must be numpy array or ExtMemArray"
+
+        if self._separator is None:
+            max_dim = feats.shape[1]
+        else:
+            assert len(feats.shape) == 1 or feats.shape[1] == 1, \
+                "When a separator is given, the input feats must be a list of strings."
+
+
+            if feats.dtype.type is not np.str_:
+                logging.warning("When a separator is given, "
+                                "the input feats must be a list of strings."
+                                "But we get %s. Will convert the input feats into str array",
+                                feats.dtype.type)
+            feats = feats.astype(str)
+            max_dim = 0
+            for feat in feats:
+                dim_size = len(feat.split(self._separator))
+                max_dim = dim_size if dim_size > max_dim else max_dim
+        return {self.feat_name: max_dim}
+
+    def update_info(self, info):
+        max_dim = max(info)
+        self._max_dim = max_dim
+
+    def call(self, feats):
+        """ Generate hard negatives as features
+
+        Parameters
+        ----------
+        feats : np array
+            Data with hard negatives.
+
+        Returns
+        -------
+        np.array
+        """
+        assert self._target_ntype is not None, \
+            "The target node type should be set, it can be the source node type " \
+            "or the destination node type depending on the hard negative case."
+        nid_map = self._nid_map[self._target_ntype]
+
+        neg_ids = np.full((len(feats), self._max_dim), -1, dtype=np.int64)
+        if self._separator is None:
+            for i, raw_ids in enumerate(feats):
```

Comment:
it's kind of confusing here. a row of `feats` is a list of ids, separated by some delimitor?